### ISSUE TITLE
Variable !BOX! not being recognized.

### ISSUE THREAD

**Issue by PostholerCom** [MapServer/MapServer#6601]
I've installed 8rc1 and I've run into an issue. I didn't see anything in the change log, so I'll put it here.

The !BOX! variable that has been historically available doesn't seem to be supported with CONNECTIONTYPE POSTGIS any longer. For instance:

```
   DATA "geom from (
      select
         geom
      from sometable
      where st_intersects(geom, !BOX!)
   ) as sq using unique geom using srid=4269"
```

Even if I comment the line out I get an error, such as:
--where st_intersects(geom, !BOX!)

If it exists any where in the DATA block, it fails.

Here's the ms error log:

[Thu Aug 25 13:46:55 2022].451955 msPostGISBuildSQL(): General error message. Failed to build SQL 'where'.
[Thu Aug 25 13:46:55 2022].451980 msPostGISLayerWhichShapes(): Query error. Failed to build query SQL.
[Thu Aug 25 13:46:55 2022].452010 msDrawMap(): Image handling error. Failed to draw layer named 'wptByTrailType'.

...http response:

<?xml version='1.0' encoding="UTF-8" standalone="no" ?>
<!DOCTYPE ServiceExceptionReport SYSTEM ["http://schemas.opengis.net/wms/1.1.1/exception_1_1_1.dtd"](http://schemas.opengis.net/wms/1.1.1/exception_1_1_1.dtd)>
<ServiceExceptionReport version="1.1.1">
<ServiceException>
msDrawMap(): Image handling error. Failed to draw layer named &#39;wptsAll&#39;.
msPostGISLayerWhichShapes(): Query error. Failed to build query SQL.
msPostGISBuildSQL(): General error message. Failed to build SQL &#39;where&#39;.
</ServiceException>
</ServiceExceptionReport>

If I remove the !BOX! it will work.

Thanks!
Scott

**Comment by jmckenna** [comment]
@PostholerCom the trick for debugging PG issues is always getting the full query sent from MapServer to PostGIS (once you get that full query, then you can test it with the `psql` commandline.  You can follow the precise steps for this at https://mapserver.org/optimization/vector.html#debugging-speed-issues-with-postgis (I had tried to slowly explain my steps there)

Or, you can also just use the `map2img` command with the `-all_debug 5` switch, to get the full msPostGISLayerWhichShapes query.  (this is what Steve was asking for in the email thread)

**Comment by PostholerCom** [comment]
That's the problem, according to the error message, it's not able to create the SQL query. Thus, no SQL is created to test with.

```
[Fri Aug 26 14:08:49 2022].411606 msWMSLoadGetMapParams(): enabling non-square pixels.
[Fri Aug 26 14:08:49 2022].414296 msDrawMap(): kicking into non-square pixel preserving mode.
[Fri Aug 26 14:08:49 2022].414416 msDrawMap(): rendering using outputformat named png (AGG/PNG).
[Fri Aug 26 14:08:49 2022].414432 msDrawMap(): WMS/WFS set-up and query, 0.000s
[Fri Aug 26 14:08:49 2022].428687 msPostGISBuildSQL(): General error message. **Failed to build SQL** 'where'.
[Fri Aug 26 14:08:49 2022].428698 msPostGISLayerWhichShapes(): Query error. **Failed to build query SQL**.
[Fri Aug 26 14:08:49 2022].428740 msDrawMap(): Image handling error. Failed to draw layer named 'nhdwaterbody'.
[Fri Aug 26 14:08:49 2022].429197 mapserv request processing time (msLoadMap not incl.): 0.018s
[Fri Aug 26 14:08:49 2022].429209 msFreeMap(): freeing map at 0x132f7b0.
```

**Comment by jmckenna** [comment]
ah true good point

**Comment by jratike80** [comment]
Does it work if `!BOX!` does not appear at all in your DATA?

`DATA "geom from sometable using unique geom using srid=4269"`

Using geom an unique value feels odd to me but I suppose you have some reason for that and it has worked for you.

**Comment by PostholerCom** [comment]
Yes, it work without the !BOX!. The parser doesn't seem to like it.

**Comment by geographika** [comment]
@PostholerCom - does your LAYER have a PROJECTION block? Also how are you calling MapServer - via a WMS/WFS? It sounds like MapServer can't calculate the BBOX value to make the replacement.

**Comment by PostholerCom** [comment]
The MAP block has a PROJECTION block, not any of the LAYER blocks.

The !BOX! variable has worked fine on every version of MapServer I've 
used since 2012, except version 8 rc1.

**Comment by jratike80** [comment]
It could make things forward much faster if you added a complete mapfile with a small dataset that others could put into PostGIS with ogr2ogr.

> All comments should be ignored.

I wonder if the comments in mapfile are analyzed at all in that place in between the double quotes of the `DATA "..."`. I recommend to forget the comments within DATA and just delete the text for now. That detail is not relevant to the !BOX! issue.

**Comment by PostholerCom** [comment]
> It could make things forward much faster if you added a complete mapfile with a small dataset that others could put into PostGIS with ogr2ogr.

If you have version 8 rc1 installed on your system, just add:
/* !BOX! */
...to a DATA block, with CONNECTIONTYPE POSTGIS.

It's very, very easy to duplicate.

**Comment by jratike80** [comment]
I agree that that part of the task is easy for me at the moment because I do have some ready mapfiles using PostGIS which I can edit. Generally I tend to be rather lazy. The more work you do for helping others to take a grip on the issue the more likely they will do that.  This time what prevents me from making a proper test is that I am on Windows and I do not have version 8 RC1 yet.

Meanwhile I tried what happens with Mapserver 7.7.0 by adding `/* !BOX! */` into DATA block as here. Without this addition this WFS layer works fine.

```
CONNECTIONTYPE POSTGIS
        CONNECTION "host=host port=5432 dbname=db user=user password=pw"
        DATA "geom from rakennus /* !BOX! */ using SRID=3067"
```

The comment within the DATA string breaks the layer so that is not anything new to version 8 RC1.
Error in this case was `msPostGISParseData(): Query error. Error parsing PostGIS DATA variable. You must specify 'using unique' when supplying a subselect in the data definition.`.

Adding `using unique geom` yields `msWFSGetFeature(): WFS server error. ms_error->code not found msPostGISLayerWhichShapes(): Query error. Error executing query. Check server logs msPostGISLayerGetExtent(): General error message. Error executing SQL. Check server logs.`

 But you are right that there is something special in `!BOX!` because this DATA does not throw errors

        DATA "geom from rakennus /* !foo! */ using SRID=3067 using unique geom"

Anyway even the foo comment gets sent forward and it is changing something in the code path so that using unique is now compulsory while in the original DATA row it is not. I do not think that using comments within DATA is a good idea. I would comment the whole old DATA row or block and write a new copy.

Disclaimer: I am a Mapserver user just like you, not developer.

**Comment by PostholerCom** [comment]
I liberally comment my DATA blocks as it's a good habit. With that said, I use '--' not /**/ that I've been touting. I should have never introduced it here.

To clarify and summarize, with CONNECTIONTYPE POSTGIS, DATA block in mapserver version 8 rc1:
   !BOX!
will fail, even if it appears in a comment such as:
   -- some comment !BOX!

**Comment by jratike80** [comment]
Hmm, I guess that the `--` comment inside data is/should be sent directly to database because it is only SQL comment https://w3schools.com/Sql/sql_comments.asp but not a mapfile comment https://www.mapserver.org/mapfile/. The ` /**/` syntax is used both by mapfile and SQL. Maybe it is somewhat undefined what will happen when that syntax appears within the SQL part of the data block.

What I can confirm is that with 7.7.0 the `--` comment with `!BBOX!` works
```
# DATA "geom from rakennus /* !foo! */ using SRID=3067"
   DATA "geom from (
	      select
	         geom, fid
	         -- comment !BBOX!
	      from rakennus
   ) as sq using srid=3067 using unique fid"
```
I may do a new test once I have managed to install Mapserver 8.

But this discussion about comments was a sidepath and your real problem is that !BBOX! does not work correctly.
I would try to set also the LAYER - PROJECTION  and LAYER - EXTENT, perhaps even LAYER - METADATA - wfs_extent and see it those helps. It is a very, very easy test. I also believe that your real DATA block is more complicated because in DATA like this using !BBOX! should not give any benefit compared to the default behavior.
```
 DATA "geom from (
      select
         geom
      from sometable
      where st_intersects(geom, !BOX!)
   ) as sq using unique geom using srid=4269"
```

**Comment by geographika** [comment]
I'm unable to recreate using `main` branch. The following:

```
DATA "geometry from (select * from osm_buildings oa where st_intersects(geom, !BOX!) ) as subquery using unique id using srid=3857"
```

Gets replaced with `"(select * from osm_buildings oa where st_intersects(geom, ST_GeomFromText('POINT(0 0)',3857)) ) as subquery"`

The error encountered is caused by being unable to recreate a valid where clause:

https://github.com/MapServer/MapServer/blob/b2f0fd240a90b003b84a9b7133f34d2d31536188/mappostgis.cpp#L2093-L2101

Adding `DEBUG 5` to the layer would likely provide the reason why this is happening. 
As @jratike80 mentioned, the example `DATA` clause should not need to add `!BBOX!` in the first place - this is only required within subqueries.

**Comment by geographika** [comment]
> Using geom an unique value feels odd to me but I suppose you have some reason for that and it has worked for you.

Well spotted @jratike80 - I can recreate the same error `"Failed to build SQL 'where'` using a geometry field as a unique identifier, so I believe this is the real cause of the issue. A proper identifier field should be used here.

**Comment by jratike80** [comment]
Unfortunately I fear that's not true. See my updated comment above. I use a proper attribute "fid" as unique but Mapserver creates lunatic SQL that ends `as sq WHERE  LIMIT 101`. I do not know what is the exact Mapserver version in Jeffs MS4W 4.0.5 that I am running.

**Comment by geographika** [comment]
That SQL is actually valid. It will always return 0 rows, but I think it is a first pass to get the field names. 

https://github.com/MapServer/MapServer/blob/b2f0fd240a90b003b84a9b7133f34d2d31536188/mappostgis.cpp#L3061

Code relating to the WHERE clause was modified as part of the move to C++, and even though it was 3 years ago it was not part of the 7.6 releases, so there well may be an issue, but it really needs a recreatable test case to debug.

**Comment by geographika** [comment]
Ignore the reference to the UID. If '!BBOX!` is used in the `DATA` statement then the `msPostGISBuildSQLWhere` doesn't append the spatial WHERE clause. If no other WHERE clauses are added (e.g. through paging, filters, selections by Id) then `msPostGISBuildSQLWhere` returns an empty string and the `Failed to build SQL 'where'` is thrown. 

In 7.6 when C was used the function returned the following in this case:

 `https://github.com/MapServer/MapServer/blob/branch-7-6/mappostgis.c#L2230

In 8.0 it returns `std::string strWhere;`

If this worked previously, then the first must have evaluated to `true` (although this doesn't seem like it should be the case according to [this link](https://stackoverflow.com/questions/36950552/0-evaluates-false-0-evaluates-true).

**Comment by jratike80** [comment]
> Ignore the reference to the UID. If !BBOX!

It is actually !BOX! but I suppose that you know it. I didn't so I spent some time with generating not related errors.

**Comment by jmckenna** [comment]
Fixed through #6602 .  Closing.

**Comment by PostholerCom** [comment]
I liberally comment my DATA blocks as it's a good habit. With that said, 
I use '--' not /**/ that I've been touting. I should have never 
introduced it here.

To clarify and summarize, with CONNECTIONTYPE POSTGIS, DATA block in 
mapserver version 8 rc1:
    !BOX!
will fail, even if it appears in a comment such as:
    -- some comment !BOX!
